bin/sessionmirror.py

```diff
@@ -0,0 +1,749 @@
+    elif src_type == 'xnat:petScanData':
+        copy_attrs(src_obj, dest_obj, PET_SCAN_ATTRS)
+    elif src_type == 'xnat:ctScanData':
+        copy_attrs(src_obj, dest_obj, CT_SCAN_ATTRS)
+    elif src_type == 'xnat:scScanData':
+        copy_attrs(src_obj, dest_obj, SC_SCAN_ATTRS)
+    elif src_type == 'proc:genProcData':
+        copy_attrs(src_obj, dest_obj, PROC_ATTRS)
+    elif src_type == 'xnat:otherDicomScanData':
+        copy_attrs(src_obj, dest_obj, OTHER_DICOM_SCAN_ATTRS)
+    else:
+        print('ERROR:cannot copy attributes, unsupported datatype:' + src_type)
+
+
+def subj_compare(item1, item2):
+    '''Compare sort of items'''
+    return cmp(item1.label(), item2.label())
+
+
+def copy_file(src_f, dest_r, cache_d):
+    '''
+    Copy file from XNAT file source to XNAT resource destination,
+    using local cache in between'''
+    f_label = src_f.label()
+    loc_f = cache_d + '/' + f_label
+
+    # Make subdirectories
+    loc_d = op.dirname(loc_f)
+    if not op.exists(loc_d):
+        os.makedirs(loc_d)
+
+    try:
+        # Download file
+        if op.exists(loc_f) is False:
+            src_f.get(loc_f)
+
+        # Get File Attributes
+        f_in_attrs = src_f.attributes()
+        f_content = f_in_attrs.get('file_content')
+        f_format = f_in_attrs.get('file_format')
+        f_tags = f_in_attrs.get('file_tags')
+
+        # Upload File
+        if f_format and f_content and not f_tags:         # format & content
+            dest_r.file(f_label).put(loc_f, f_format, f_content)
+        elif f_format and not f_content and not f_tags:   # format only
+            dest_r.file(f_label).put(loc_f, f_format)
+        elif f_format and f_content and f_tags:   # format, content, & tags
+            dest_r.file(f_label).put(loc_f, f_format, f_content)
+        else:                                             # none
+            dest_r.file(f_label).put(loc_f)
+
+        # Delete local copy
+        os.remove(loc_f)
+    except Exception:
+        print("ERROR:failed to copy file:%s, error=%s"
+              % (f_label, sys.exc_info()[0]))
+
+
+def copy_res_zip(src_r, dest_r, cache_d):
+    '''
+    Copy a resource from XNAT source to XNAT destination using local cache
+    in between
+    '''
+    try:
+        # Download zip of resource
+        print('INFO:Downloading resource as zip...')
+        cache_z = src_r.get(cache_d, extract=False)
+
+        # Upload zip of resource
+        print('INFO:Uploading resource as zip...')
+        dest_r.put_zip(cache_z, extract=True)
+
+        # Delete cached zip
+        os.remove(cache_z)
+
+    except IndexError:
+        print('ERROR:failed to copy:%s:%s' % (cache_z, sys.exc_info()[0]))
+        raise
+
+
+def is_empty_resource(_res):
+    '''Check if resource contains any files'''
+    f_count = 0
+    for f_in in _res.files().fetchall('obj'):
+        f_count += 1
+        break
+
+    return f_count == 0
+
+
+def copy_project(src_proj, dst_proj, proj_cache_dir):
```

Comment:
Note that `copy_project` function does not copy project-level resources.

Reply:
Right. Actually this function was just copied/pasted. This is not tested nor used by `sessionmirror.py`, as this only copies sessions (the original `xnat_mirror.py` is intended for entire projects). But you are right, project-level resources seem to be ignored. We may also just remove them for the moment.